libs/remote_storage/src/s3_bucket.rs

```diff
@@ -485,89 +484,93 @@ impl RemoteStorage for S3Bucket {
                 })
             });
 
-        let _permit = self.permit(kind, cancel).await?;
-
-        let mut continuation_token = None;
-
-        loop {
-            let started_at = start_measuring_requests(kind);
-
-            // min of two Options, returning Some if one is value and another is
-            // None (None is smaller than anything, so plain min doesn't work).
-            let request_max_keys = self
-                .max_keys_per_list_response
-                .into_iter()
-                .chain(max_keys.into_iter())
-                .min();
-            let mut request = self
-                .client
-                .list_objects_v2()
-                .bucket(self.bucket_name.clone())
-                .set_prefix(list_prefix.clone())
-                .set_continuation_token(continuation_token)
-                .set_max_keys(request_max_keys);
-
-            if let ListingMode::WithDelimiter = mode {
-                request = request.delimiter(REMOTE_STORAGE_PREFIX_SEPARATOR.to_string());
-            }
-
-            let request = request.send();
-
-            let response = tokio::select! {
-                res = request => res,
-                _ = tokio::time::sleep(self.timeout) => return Err(DownloadError::Timeout),
-                _ = cancel.cancelled() => return Err(DownloadError::Cancelled),
-            };
-
-            let response = response
-                .context("Failed to list S3 prefixes")
-                .map_err(DownloadError::Other);
-
-            let started_at = ScopeGuard::into_inner(started_at);
-
-            crate::metrics::BUCKET_METRICS
-                .req_seconds
-                .observe_elapsed(kind, &response, started_at);
-
-            let response = response?;
-
-            let keys = response.contents();
-            let empty = Vec::new();
-            let prefixes = response.common_prefixes.as_ref().unwrap_or(&empty);
-
-            tracing::debug!("list: {} prefixes, {} keys", prefixes.len(), keys.len());
+        async_stream::stream! {
+            let _permit = self.permit(kind, cancel).await?;
+
+            let mut continuation_token = None;
+            'outer: loop {
+                let started_at = start_measuring_requests(kind);
+
+                // min of two Options, returning Some if one is value and another is
+                // None (None is smaller than anything, so plain min doesn't work).
+                let request_max_keys = self
+                    .max_keys_per_list_response
+                    .into_iter()
+                    .chain(max_keys.into_iter())
+                    .min();
+                let mut request = self
+                    .client
+                    .list_objects_v2()
+                    .bucket(self.bucket_name.clone())
+                    .set_prefix(list_prefix.clone())
+                    .set_continuation_token(continuation_token)
+                    .set_max_keys(request_max_keys);
```

Comment:
Did you check out https://docs.rs/aws-sdk-s3/1.26.0/aws_sdk_s3/operation/list_objects_v2/builders/struct.ListObjectsV2FluentBuilder.html#method.into_paginator and decided not to use it?

Reply:
If we used that function, we'd probably rely on the SDK provided retry behaviour. but we want to do retries ourselves, and explicitly disable SDK provided retries.

Reply:
Hmm.... Aah yes, I didn't think of retries at all. Shouldn't we have them here, otherwise one would have to re-start the stream from the beginning?

Reply:
To get on with this PR and punt it, it only needs to be considered with the scrubber. Maybe add something to he documentation about this hazard?

Alternatively, we could rip out/abstract the continuation token somehow.

Reply:
Yeah either we'd have to move retries into the list_streaming call, or return a `ListingRequest` object in the Stream, that then would have a retryable `async fn list()` function implemented.

Personally I'd do retries in the next step, this first PR was to have a streaming API at all. Before that I wanted to check how azure handles retries.

Reply:
Handled on a follow-up.